### PR TITLE
feat(shared-data): WIP pydantic model generation

### PR DESCRIPTION
**Pull request by amitlissack**
# Overview

This is not for real. It's just a place to see how model generation might work. 

Or .. for those not familiar with pydantic, how models might look from which we could generate a schema.

# Changelog

- Modified the protocol schema to make it friendlier for code generation. Anonymous objects (like the commands) were given names like `Command1`, `Command2`, etc.
- Add rubbish shell script that uses `datamodel-code-generator` to generate pydantic models from labware and protocol jsonnschema files.
- Added `generated.json` which is the protocol schema that Pydantic generates from model that `datamodel-code-generator` generated. Generate!!

# Review requests

# Risk assessment

None. This will not be merged.

### CODE REVIEW

File: shared-data/protocol/schemas/5.json [hunk 1]
```diff
@@ -170,527 +644,75 @@
...
-            "mount": {
-              "description": "Where the pipette is mounted",
-              "type": "string",
-              "enum": ["left", "right"]
-            },
-            "name": {
-              "$ref": "#/definitions/pipetteName"
-            }
+      "type": "object",
+      "additionalProperties": {
```
Comment (amitlissack) [inline shared-data/protocol/schemas/5.json:648]: The generator does not like `patternProperties`. This is not meaningful in this instance.

File: shared-data/protocol/schemas/5.json [hunk 2]
```diff
@@ -170,527 +644,75 @@
...
         }
       }
     },
 
     "labwareDefinitions": {
       "description": "All labware definitions used by labware in this protocol, keyed by UUID",
-      "patternProperties": {
-        ".+": {
+      "type": "object",
+      "additionalProperties": {
```
Comment (amitlissack) [inline shared-data/protocol/schemas/5.json:669]: The generator does not like `patternProperties`. This is not meaningful in this instance.

File: shared-data/protocol/schemas/5.json [hunk 3]
```diff
@@ -170,527 +644,75 @@
...
-            "definitionId": {
-              "description": "reference to this labware's ID in \"labwareDefinitions\"",
-              "type": "string"
-            },
-            "displayName": {
-              "description": "An optional human-readable nickname for this labware. Eg \"Buffer Trough\"",
-              "type": "string"
-            }
+      "type": "object",
+      "additionalProperties": {
```
Comment (amitlissack) [inline shared-data/protocol/schemas/5.json:677]: The generator does not like `patternProperties`. This is not meaningful in this instance.

File: shared-data/protocol/schemas/5.json [hunk 4]
```diff
@@ -170,527 +644,75 @@
...
-          "required": ["slot", "model"],
-          "additionalProperties": false,
-          "properties": {
-            "slot": { "$ref": "#/definitions/slot" },
-            "model": {
-              "description": "model of module. Eg 'magneticModuleV1' or 'magneticModuleV2'. This should match a top-level key in shared-data/module/definitions/2.json",
-              "type": "string"
-            }
+      "type": "object",
+      "additionalProperties": {
```
Comment (amitlissack) [inline shared-data/protocol/schemas/5.json:699]: The generator does not like `patternProperties`. This is not meaningful in this instance.

File: shared-data/python/opentrons_shared_data/generated/opentronsLabwareSchemaV2.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    ordering: List[OrderingItem] = Field(
+        ...,
+        description='Generated array that keeps track of how wells should be ordered in a labware',
+    )
+    cornerOffsetFromSlot: CornerOffsetFromSlot = Field(
+        ...,
+        description='Distance from left-front-bottom corner of slot to left-front-bottom corner of labware bounding box. Used for labware that spans multiple slots. For labware that does not span multiple slots, x/y/z should all be zero.',
+    )
+    dimensions: Dimensions = Field(..., description='Outer dimensions of a labware')
+    wells: Dict[str, Any] = Field(
```
Comment (amitlissack) [inline shared-data/python/opentrons_shared_data/generated/opentronsLabwareSchemaV2.py:183]: The generator does not like `patternProperties`. And it is meaningful in this instance!

The json schema uses a regular expression to validate the well name. A significant limitation of the code generator.

Comment (mcous) [reply shared-data/python/opentrons_shared_data/generated/opentronsLabwareSchemaV2.py:183]: Per #7531, I wonder if we should re-consider our constraints on the well name. I believe removing this constraint would be a loosening of the spec, and therefore not a breaking change.

I can't really think of a good reason why a user shouldn't be able to name their wells whatever they want (acknowledging that the current [letter][number] scheme is currently the only way a user can have the concept of a "row name" and a "column name")

Comment (amitlissack) [reply shared-data/python/opentrons_shared_data/generated/opentronsLabwareSchemaV2.py:183]: I'm with you. It's interesting to note that we don't enforce well naming in the `ordering` member.

Comment (Laura-Danielle) [reply shared-data/python/opentrons_shared_data/generated/opentronsLabwareSchemaV2.py:183]: To be honest, that non-enforcement in `ordering` was probably an oversight. I would be curious to know if users actually care to name their wells anything other than [letter][number] which is in line with how most labware is named.

I am fine with loosening the specs regardless, though.

**Comment by Laura-Danielle** [comment]
This is pretty cool! Would the idea behind this be to take in PD protocols and then generate pydantic commands for use in the protocol engine, or is there some other purpose?

**Comment by amitlissack** [comment]
> This is pretty cool! Would the idea behind this be to take in PD protocols and then generate pydantic commands for use in the protocol engine, or is there some other purpose?

Yes. 

The models could replace the typedict based models in python shared data.

**Comment by mattwelch** [comment]
Marked as accepted. **Don't delete the branch.**

**Pull request closed without merging**